### PR TITLE
Fix some tooling for windows

### PR DESCRIPTION
**Pull request by jplu**
# What does this PR do?
This PR fixes a small issue on the `check_repo.py` script with the encodings. I have also updated the `Makefile` to move the creation of the variables `fork_point_sha`, `joined_dirs` and `modified_py_files` in the `modified_only_fixup` target because they are used only there.

The makefile raises an error on Windows environment when creating the `modified_py_files` variable:
```
'tests' is not recognized as an internal or external command, operable program or batch file
```

ping @stas00 and @sgugger

### CODE REVIEW

**Review by sgugger — APPROVED** [review]
LGTM for the changes in `check_repo` (those are things I forgot, sorry!). Thanks for fixing!
For the MakeFile, I'll defer to @stas00

File: Makefile
Comment (stas00) [inline Makefile:7]: Unfortunately, this doesn't work. Makefile handles things differently inside targets - i.e. none of this code will work after this move.

I was hoping to avoid this as inside the target it calls for a much more cumbersome code, so unless it's in the way I'd rather have it stay where it is. If it is a problem then this will have to be re-worked.

Comment (stas00) [reply Makefile:7]: I do agree though, that ideally that piece of code shouldn't execute unless it's needed. I just saw that it was very quick and caused no interference.

Comment (stas00) [reply Makefile:7]: If it's not in a way, but something we want to improve please make a separate issue and tag me and I will tackle it at a later time. Thanks.

Comment (sgugger) [reply Makefile:7]: The problem is that the current MakeFile does not work for all windows users, preventing them from running the necessary commands to clean their code.

Comment (sgugger) [reply Makefile:7]: (So we do need a fix.)

Comment (jplu) [reply Makefile:7]: Indeed, I have to run the make in my WSL terminal to make it work otherwise it fails with the error I gave.

As @stas00 propsed I can revert that part and open an issue that you can check afterwards, in order to at least push the other fix. Is that ok for you @sgugger ?

Comment (stas00) [reply Makefile:7]: Ah, OK, the OP doesn't say that the proposed change is to fix a problem.  Thank you for clarifying that it's an actual problem.

I will work on it in a few hours and make it work inside the target.

Also I don't know what the problem is on windows. Perhaps it's a much simpler fix to make it work on windows rather than hide, as they won't be able to use this optimized target anyway it seems. I hope it make sense.

Comment (sgugger) [reply Makefile:7]: Reverting that PR to merge the fixes in check_repo is a nice first step, then opening an issue would be nice. Could you copy the full error message for @stas00 in it @jplu ?

Comment (stas00) [reply Makefile:7]: If it's not urgent to merge right away please give me some time and then there is no need to open another issue. If you're in a hurry then do whatever needs to be done and meanwhile I will work on making it work in the target.

If you want an actual fix so that the optimized target works on windows too please communicate the errors you encountered - most likely these have to do with the type of shell you run on windows and probably not having git and some of the unix tools, but this is a wild guess.

Comment (jplu) [reply Makefile:7]: We are not in a hurry, no rush, we can wait :)

The error message is:
```
'tests' is not recognized as an internal or external command, operable program or batch file
```

What I have tracked is that the `egrep` command from the GNU CoreUtils for Windows doesn't interpret properly the following part of the command: `egrep '^($(joined_dirs))'`

Comment (stas00) [reply Makefile:7]: Great! 

any difference if you use `grep -E` instead of `egrep`?

if not if you replace   `egrep '^($(joined_dirs))'` with a hardcoded: egrep '(examples|tests|src|utils)'`

Ideally if you know your shell, and could make it work there - it'd be much easier to integrate with the unix version. 

If all fails I'd probably switch to a python built-in program, which should be more cross-platform - so will only need to rely on shell+git commands to work.

Comment (jplu) [reply Makefile:7]: All fails with the same error :( and `make -n` gives me:

```
process_begin: CreateProcess(NULL, "", ...) failed.
process_begin: CreateProcess(NULL, "", ...) failed.
if [ -n "" ]; then \
                echo "Checking/fixing "; \
                black ; \
                isort ; \
                flake8 ; \
        else \
                echo "No library .py files were modified"; \
        fi
```

Comment (stas00) [reply Makefile:7]: From looking at other reports on SO I interpret `process_begin: CreateProcess(NULL, "",` is that it invokes a shell without any command - too bad it's not showing what's hidden behind `...`.

OK, let's approach it differently. Assuming you know windows shell - how would you solve this differently - all this code is doing is trying to isolate files that were modified in a subset of folders. perhaps we could just if /windows/ this else that.

But if it's not immediately obvious please don't worry I'm going to look at rewriting it in python and then you can help me to validate that it's cross platform.

Comment (jplu) [reply Makefile:7]: Basically if I run these lines one by one:
```
fork_point_sha := $(shell git merge-base --fork-point master)
joined_dirs := $(shell echo $(check_dirs) | tr " " "|")
modified_py_files := $(shell git diff --name-only $(fork_point_sha) | egrep '^($(joined_dirs))' | egrep '\.py$$')
```

In any Windows shell (Windows Terminals, Windows PowerShell, Windows Prompt) it works like expected. Only when in the Makefile it doesn't works. So I don't really know what kind of workaround I can apply :/

Comment (stas00) [reply Makefile:7]: I'm not 100% sure it actually works since `$$` is a special thing inside Makefile, so that code can't be copy-n-pasted to the shell - so what most likely what you're say is that there is no error.

Thank you for doing this debug/diagnosis work, @jplu 

Having seen this kind of issue in the past and running into a multiple variations of windows shells, I think a much safer bet is to just rewrite this functionality in python. So as long as `git` works consistently in whatever shell is used it'll work everywhere. I will work on it shortly.

Comment (jplu) [reply Makefile:7]: Sorry, more precisely what I have done is to run each line but without the `shell` keyword and the `$$`.

> Having seen this kind of issue in the past and running into a multiple variations of windows shells, I think a much safer bet is to just rewrite this functionality in python. So as long as git works consistently in whatever shell is used it'll work everywhere. It's just going to take 10 times more code. I will work on it shortly.

Yes, I think it should be much better. Thank you very much for taking care this!! I will revert the Makefile part for this PR.

Comment (stas00) [reply Makefile:7]: Thank you for that extra clarification, @jplu. It is awesome you know of these Makefile nuances.

Perhaps Makefile calls a different shell internally on windows, hence the different behavior. See "Choosing a Shell in DOS and Windows" at https://www.gnu.org/software/make/manual/html_node/Choosing-the-Shell.html

Python it is.

Comment (jplu) [reply Makefile:7]: Ok, I set my `SHELL` variable to `C:\WINDOWS\System32\WindowsPowerShell\v1.0\powershell.exe` and I still have:
```
process_begin: CreateProcess(NULL, "", ...) failed.
process_begin: CreateProcess(NULL, "", ...) failed.
if [ -n "" ]; then \
                echo "Checking/fixing "; \
                black ; \
                isort ; \
                flake8 ; \
        else \
                echo "No library .py files were modified"; \
        fi
```
When running `make -n`

Comment (jplu) [reply Makefile:7]: Here what I have with `make -dr` :
```
GNU Make 3.81
Copyright (C) 2006 Free Software Foundation, Inc.
This is free software; see source for copying conditions.
There is NO guarantee; both for COMMERCIAL use and for
MEET A SPECIAL NEED.

This software is build for i386-pc-mingw32
Reading makefiles ...
Reading the makefile "makefile" ...
CreateProcess (C:\Users\snake\AppData\Local\GitHubDesktop\app-2.5.4\resources\app\git\mingw64\bin\git.exe, git merge-base --fork-point master, ...)
Main thread handle = 0x00000198
Creating a temporary file C:\Users\snake\AppData\Local\Temp\make11588-1.bat
CreateProcess (C:\Users\snake\AppData\Local\Temp\make11588-1.bat, C:\Users\snake\AppData\Local\Temp\make11588-1.bat, ...)
Cleaning the temporary file C:\Users\snake\AppData\Local\Temp\make11588-1.bat
Creating a temporary file C:\Users\snake\AppData\Local\Temp\make11588-1.bat
CreateProcess (C:\Users\snake\AppData\Local\Temp\make11588-1.bat, C:\Users\snake\AppData\Local\Temp\make11588-1.bat, ...)
'tests' is not recognized as an internal command
or external, an executable program or a batch file.
Cleaning the temporary file C:\Users\snake\AppData\Local\Temp\make11588-1.bat
Updated makefiles ....
 Study the target file "makefile".
  Finding an implicit rule for "makefile".
  No default rule found for "makefile".
  End of dependencies of the target "makefile" file.
 No need to remake the "makefile" target.
Update of targeted targets ....
Study of the target file "modified_only_fixup".
 The "modified_only_fixup" file does not exist.
 End of dependencies of the target file "modified_only_fixup".
We must remake the target "modified_only_fixup".
Creating a temporary file C:\Users\snake\AppData\Local\Temp\make11588-1.bat
CreateProcess (C:\Users\snake\AppData\Local\Temp\make11588-1.bat, C:\Users\snake\AppData\Local\Temp\make11588-1.bat, ...)
Added child process 0x02b1fec0 (modified_only_fixup) PID 45282048 to the chain.
Active child process 0x02b1fec0 (modified_only_fixup) PID 45282048
-n was unexpected.
Retrieving the exit status of the child process 0x02b1fec0 PID 45282048
Cleaning the temporary file C:\Users\snake\AppData\Local\Temp\make11588-1.bat
make: *** [modified_only_fixup] Error 255
Removed child process 0x02b1fec0 PID 45282048 from the chain.
```

Comment (stas00) [reply Makefile:7]: Here we go: https://github.com/huggingface/transformers/pull/8371

Comment (stas00) [reply Makefile:7]: > -n was unexpected.

Is it referring to?
```
@if [ -n "$(modified_py_files)" ]; then \
```
So is it possible it can't handle `-n` - so switching to python still won't work. then. Let's continue debugging this in https://github.com/huggingface/transformers/pull/8371

Comment (jplu) [reply Makefile:7]: I think as this part is written in shell, it cannot be properly handled in any Windows shell. But this is my own interpretation.

**Pull request merged**
